### PR TITLE
[BTS-800] Cluster not enabled exception

### PR DESCRIPTION
**Pull request by apetenchea**
### Scope & Purpose

This aims to fix the following error: `caught exception during start of feature 'FailureOracle': feature 'Cluster' is not enabled (exception location: /work/ArangoDB/lib/ApplicationFeatures/ApplicationServer.h:466)`.

It is suspected that during `UpgradeFeature` the `ClusterFeature` gets disabled, and thus other features which depend on it should too.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by dothebart — APPROVED** [review]

File: arangod/RestServer/UpgradeFeature.cpp [hunk 1]
```diff
@@ -148,6 +149,10 @@ void UpgradeFeature::validateOptions(std::shared_ptr<ProgramOptions> options) {
       server().getFeature<ReplicationFeature>();
   replicationFeature.disableReplicationApplier();
 
+  cluster::FailureOracleFeature& failureOracleFeature =
```
Comment (mchacki) [inline arangod/RestServer/UpgradeFeature.cpp:152]: Can you please move this into the list of `_nonServerFeatures` that is handed in from the outside.This way we have it at least in the same position as the other force-disabled features.

File: arangod/Cluster/FailureOracleFeature.cpp [hunk 2]
```diff
@@ -213,7 +215,7 @@ void FailureOracleFeature::prepare() {
 void FailureOracleFeature::start() {
   TRI_ASSERT(_cache == nullptr);
   _cache = std::make_shared<FailureOracleImpl>(
-      server().getEnabledFeature<ClusterFeature>());
+      server().getFeature<ClusterFeature>());
```
Comment (mchacki) [inline arangod/Cluster/FailureOracleFeature.cpp:218]: This change should not be necessary with any of the other two changes.

File: arangod/Cluster/FailureOracleFeature.cpp [hunk 3]
```diff
@@ -202,6 +202,8 @@ FailureOracleFeature::FailureOracleFeature(Server& server)
   setOptional(true);
   startsAfter<SchedulerFeature>();
   startsAfter<ClusterFeature>();
+  onlyEnabledWith<SchedulerFeature>();
+  onlyEnabledWith<ClusterFeature>();
```
Comment (mchacki) [inline arangod/Cluster/FailureOracleFeature.cpp:206]: In my opinion this is the desired fix and should be the only one in this PR.

File: arangod/RestServer/UpgradeFeature.cpp (hunk 1 above)
Comment (mchacki) [reply arangod/RestServer/UpgradeFeature.cpp:152]: I would actually prefer to not have this, but only have the "onlyEnabledWith<>".

**Review by mchacki — CHANGES_REQUESTED** [review]
In this PR we have 3 additions to work around the same problem.

One is a workaround by just using "getFeature" instead of "getEnabledFeature" I would like to not have this workaround, as the other two should each fix the situation properly.

The onlyEnabledWith<> is my preferred solution, as it tells our dependency tree that the features should belong together and only work together. So this is future proof even if someone disables the cluster (or scheduler) somewhere else in the future.

Force disabling the feature will work as of today, but looks like a rather manual patch on one place, I would like to not include this.

File: arangod/RestServer/UpgradeFeature.cpp (hunk 1 above)
Comment (apetenchea) [reply arangod/RestServer/UpgradeFeature.cpp]: Reverted this file back to devel version

File: arangod/Cluster/FailureOracleFeature.cpp (hunk 2 above)
Comment (apetenchea) [reply arangod/Cluster/FailureOracleFeature.cpp]: Reverted

**Review by mchacki — APPROVED** [review]
LGTM

**Pull request merged**